A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+`weighted_target_experimental` depending on the action, so we can skip the
+`xds_routing` policy that has only one child.
+
+The `xds_routing_experimental` balancer config will contain the validated
+`{matchers, action}` pairs. The actions will be in a separate list, and will
+be referred to by the routes by their names.
+
+#### Action list
+
+All the actions from the validated `{matchers, action}` pairs will be
+de-duplicated and stored in a cache. All actions in the cache will be
+assigned a unique name.
+
+When processing an RDS response, for each action:
+
+*   If the action is cluster, the child\_policy will be `cds_experimental`, with
+    a balancer config containing the cluster string.
+*   If the action is weighted\_clusters, the child\_policy will be
+    `weighted_target_experimental`. The balancer config will contain a list of
+    `cds_experimental` LB configs, and their corresponding weights.
+*   If this action is already in the cache, it will be reused, and nothing needs
+    to be done
+*   If the action is new, it will be be added to the list and assigned a name
+    *   The name should be a human readable hash of the config, to be logged for
+        debugging purposes.
+    *   For example, the names can be `cds:cluster1` and
+        `weighted:cluster1_cluster2_<index_number>`.
+
+When assigning names to new actions, we will try to reuse the existing names.
+This will help the xds\_routing policy to not always close/create child
+policy. But we will limit the name reuse only between weighted cluster
+actions that share the same set of clusters (the only update is weights).
+This covers the typical use case where the user keeps updating the cluster
+weights during a canary rollout.
+
+To do this, for each new RDS response, the client will
+
+1.  Calculate the list of new actions to add, and the list of old actions to remove.
+    1.  This is to find what actions changed between the new response and the previous response, two actions are the same if they have the same set of clusters, and the same weight for each cluster.
+1.  For each new action, either find an action to reuse, or create a new action
+    1.  If there’s a to-be-removed action with the same set of clusters (but different weights), it will get the old action’s name, otherwise it will be assigned a new name.
+1.  To accelerate processing, use 2-level-maps to keep track of the actions
+    1.  First level key is cluster names without weight, second level is cluster names + weights
+    1.  See appendix for [pseudocode](#Action-name-reusing-pseudocode).
+
+
+#### Route list
+
+For each `{matchers, action}`, a route will be generated:
+
+*   With matchers in json representation.
+*   The action will be the name of this action from the action list.
+
+The processed response will be converted to a [JSON service
+config](#JSON-service-config-generated-by-xDS-resolver).
+
+### Multiple watches for the same resource
```

Comment:
IIRC, all of the logic in this section is existing logic and not a new proposal. If true, can we reference existing doc/section?

Reply:
We don't have another doc with this behavior. This is the first time we will actually have multiple watchers for the same resource.